File: src/main/resources/scripts/py/rampup.py
from loadbuilder import init as loadBuilderInit
from timeout import init as timeOutInit
from chain import build as chainBuild
from chain import execute as chainExecute
#
from java.lang import Long, Integer, Throwable, NumberFormatException, InterruptedException
from java.util.concurrent import TimeUnit
#
from org.apache.logging.log4j import Level, LogManager, ThreadContext
#
from com.emc.mongoose.common.conf import RunTimeConfig, SizeUtil
from com.emc.mongoose.common.log import LogUtil, Markers
#
LOG = LogManager.getLogger()
#
def init():
	runTimeConfig = RunTimeConfig.getContext()
	LOG.debug(Markers.MSG, "Setting the metric update period to zero for chain scenario")
	runTimeConfig.set(RunTimeConfig.KEY_LOAD_METRICS_PERIOD_SEC, 0)
	#
	loadTypesChain = runTimeConfig.getStringArray(RunTimeConfig.KEY_SCENARIO_CHAIN_LOAD)
	LOG.info(Markers.MSG, "Load types chain: {}", loadTypesChain)
	listSizes = runTimeConfig.getStringArray(RunTimeConfig.KEY_SCENARIO_RAMPUP_SIZES)
	LOG.info(Markers.MSG, "Data sizes: {}", listSizes)
	listConnCounts = runTimeConfig.getStringArray(RunTimeConfig.KEY_SCENARIO_RAMPUP_CONN_COUNTS)
	LOG.info(Markers.MSG, "Thread counts: {}", listConnCounts)
	return loadTypesChain, listSizes, listConnCounts
#
def execute(loadBuilder, rampupParams=((),(),()), timeOut=Long.MAX_VALUE, timeUnit=TimeUnit.DAYS):
	loadTypesChain = rampupParams[0]
	listSizes = rampupParams[1]
	listConnCounts = rampupParams[2]
	flagConcurrent = False
	for index, dataItemSizeStr in enumerate(listSizes):
		dataItemSize = Long(SizeUtil.toSize(dataItemSizeStr))
		for connCountStr in listConnCounts:
			try:
				connCount = Integer.parseInt(connCountStr)
				LOG.info(Markers.PERF_SUM, "---- Step {}x{} start ----", connCount, dataItemSizeStr)
				ThreadContext.put("currentSize", dataItemSizeStr + "-" + str(index))
				ThreadContext.put("currentConnCount", str(connCount))
				nextChain = chainBuild(
					loadBuilder, loadTypesChain, flagConcurrent, True, dataItemSize, dataItemSize,
					connCount
				)
				chainExecute(nextChain, flagConcurrent, timeOut, timeUnit)
				LOG.debug(Markers.MSG, "---- Step {}x{} finish ----", connCount, dataItemSizeStr)
			except InterruptedException as e:
				raise e
			except NumberFormatException as e:
				LogUtil.exception(Markers.ERR, Level.WARN, e, "Failed to parse the next thread count")
#
if __name__ == "__builtin__":
	loadBuilder = loadBuilderInit()
	stepTime = timeOutInit()
	try:
		execute(loadBuilder=loadBuilder, rampupParams=init(), timeOut=stepTime[0], timeUnit=stepTime[1])
	except InterruptedException as e:
		LOG.debug(Markers.MSG, "Rampup was interrupted")
	except Throwable as e:
		LogUtil.exception(LOG, Level.ERROR, e, "Scenario failed")
	loadBuilder.close()
	LOG.info(Markers.MSG, "Scenario end")

File: src/main/java/com/emc/mongoose/util/scenario/Rampup.java
package com.emc.mongoose.util.scenario;
//
import com.emc.mongoose.common.conf.RunTimeConfig;
import com.emc.mongoose.common.conf.SizeUtil;
import com.emc.mongoose.common.log.LogUtil;
import com.emc.mongoose.common.log.Markers;
//
import com.emc.mongoose.core.api.load.builder.LoadBuilder;
//
import com.emc.mongoose.run.cli.HumanFriendly;
import com.emc.mongoose.util.scenario.shared.WSLoadBuilderFactory;
//
import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.ThreadContext;
//
import java.rmi.RemoteException;
import java.util.Map;
import java.util.concurrent.TimeUnit;
/**
 Created by kurila on 11.06.15.
 */
public class Rampup
implements Runnable {
	//
	private final static Logger LOG;
	static {
		LogUtil.init();
		LOG = LogManager.getLogger();
	}
	//
	private final LoadBuilder loadBuilder;
	private final long timeOut;
	private final TimeUnit timeUnit;
	private final String loadTypeSeq[], sizeSeq[], threadCountSeq[];
	//
	public Rampup(
		final LoadBuilder loadBuilder, final long timeOut, final TimeUnit timeUnit,
		final String loadTypeSeq[], final String sizeSeq[], final String threadCountSeq[]
	) {
		this.loadBuilder = loadBuilder;
		this.timeOut = timeOut;
		this.timeUnit = timeUnit;
		this.loadTypeSeq = loadTypeSeq;
		this.sizeSeq = sizeSeq;
		this.threadCountSeq = threadCountSeq;
	}
	//
	@Override
	public final void run() {
		Chain nextLoadSeq;
		for(int i = 0; i < sizeSeq.length; i++) {
			final String nextSizeStr = sizeSeq[i];
			for(final String nextThreadCountStr : threadCountSeq) {
				ThreadContext.put("currentSize", nextSizeStr + "-" + i);
				ThreadContext.put("currentThreadCount", nextThreadCountStr);
				final long nextSize = SizeUtil.toSize(nextSizeStr);
				final String nextStepName = nextThreadCountStr + "x" + nextSizeStr;
				LOG.debug(Markers.MSG, "Build the next step load chain: \"{}\"", nextStepName);
				try {
					loadBuilder
						.setMinObjSize(nextSize)
						.setMaxObjSize(nextSize)
						.setThreadsPerNodeDefault(Integer.parseInt(nextThreadCountStr));
					nextLoadSeq = new Chain(loadBuilder, timeOut, timeUnit, loadTypeSeq, false);
					LOG.info(Markers.PERF_SUM, "---- Step {} start ----", nextStepName);
					nextLoadSeq.run();
				} catch(final RemoteException e) {
					LogUtil.exception(LOG, Level.WARN, e, "Failed to apply rampup params remotely");
				}
			}
		}
	}
	//
	public static void main(final String... args) {
		try {
			//
			RunTimeConfig.initContext();
			final RunTimeConfig runTimeConfig = RunTimeConfig.getContext();
			// load the config from CLI arguments
			final Map<String, String> properties = HumanFriendly.parseCli(args);
			if(!properties.isEmpty()) {
				LOG.debug(Markers.MSG, "Overriding properties {}", properties);
				RunTimeConfig.getContext().overrideSystemProperties(properties);
			}
			//
			LOG.info(Markers.MSG, RunTimeConfig.getContext().toString());
			//
			final LoadBuilder loadBuilder = WSLoadBuilderFactory.getInstance(runTimeConfig);
			final long timeOut = runTimeConfig.getLoadLimitTimeValue();
			final TimeUnit timeUnit = runTimeConfig.getLoadLimitTimeUnit();
			// adjust some defaults if necessary
			LOG.debug(Markers.MSG, "Setting the metric update period to zero for chain scenario");
			runTimeConfig.set(RunTimeConfig.KEY_LOAD_METRICS_PERIOD_SEC, 0);
			//
			final String[] loadTypeSeq = runTimeConfig.getScenarioChainLoad();
			final String[] sizeSeq = runTimeConfig.getScenarioRampupSizes();
			final String[] threadCountSeq = runTimeConfig.getScenarioRampupThreadCounts();
			final Rampup rampupScenario = new Rampup(
				loadBuilder, timeOut, timeUnit, loadTypeSeq, sizeSeq, threadCountSeq
			);
			//
			rampupScenario.run();
		} catch(final Exception e) {
			e.printStackTrace(System.err);
			LogUtil.exception(LOG, Level.ERROR, e, "Scenario failed");
		}
	}
}
